### PR TITLE
Support of transformation parameters for the interaction box

### PR DESCRIPTION
**Pull request by MBPhys**
# Description
Until now, the interaction box did not adopt the layer transformation parameters like scaling, translation, ... ( see the attached videos). 

https://user-images.githubusercontent.com/83079938/159717943-7a1c78cc-a3d8-4f10-9493-f625bb460c90.mp4

For use cases where these parameters are different from the standard values (e.g. defined in a layers metadata when read from file), the interaction box does not fit the actual layer.
Therefore we solved this, see the attached videos.

https://user-images.githubusercontent.com/83079938/159717956-eedc2604-fc31-4025-954a-2fb32315c9b9.mp4

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/components/_interaction_box_mouse_bindings.py
Comment (alisterburt) [inline napari/components/_interaction_box_mouse_bindings.py]: minor nitpick: `ini_transf` -> `initial_transform`

is this just the initial transform or is it more generally the layer transform? if so, you could call it `layer_transform`

because the inverse is available at `self._ini_transf.inverse` I'm not sure we need to store the inverse transform directly here

File: napari/components/_interaction_box_mouse_bindings.py
Comment (alisterburt) [inline napari/components/_interaction_box_mouse_bindings.py]: general question not specific to this PR, I see these events get connected but not disconnected, is this handled anywhere?

File: napari/components/_interaction_box_mouse_bindings.py
Comment (alisterburt) [inline napari/components/_interaction_box_mouse_bindings.py]: as above, don't need to store the inverse explicitly

File: napari/components/_interaction_box_mouse_bindings.py
Comment (alisterburt) [inline napari/components/_interaction_box_mouse_bindings.py]: for readability here I would separate this out to something like

```
current_affine = self._ref_viewer().layers.selection.active.affine
layer_transform = event.value.compose(self._ini_trans.inverse)
current_affine.replace_slice(layer_dims_displayed, layer_transform)
```

what do you think?

**Review by alisterburt — COMMENTED** [review]
Thanks for the PR @MBPhys ! This is awesome :) I've left a couple of minor comments I'd like to see addressed then I see no reason this can't be merged!

maybe @kevinyamauchi or @jni would also like to weigh in?

File: napari/components/_interaction_box_mouse_bindings.py
Comment (MBPhys) [reply napari/components/_interaction_box_mouse_bindings.py]: First of all, thanks for the quick review :)
I have changed this to initial_transform

It is rather a part of the layer transform without the affine part, an initial transform of scaling, translation, so on. Therefore the name.

I store and build the inverse transform one time because otherwise the inverse is computed every time when the affine transform changes. I thought this should be not the case and can save computation time in this way.
Of course, the difference will not be huge. 

https://github.com/napari/napari/blob/906f772c01d6a4cb33d5d957f3cfd452b068145f/napari/utils/transforms/transforms.py#L504-L506

File: napari/components/_interaction_box_mouse_bindings.py
Comment (MBPhys) [reply napari/components/_interaction_box_mouse_bindings.py]: Thanks:)

I have removed this in this function and put this into the layer_affine_event_helper and add the disconnection there, too:)

https://github.com/napari/napari/blob/6352c2f634a5ecefb944f2b68395adaf63b54a0f/napari/components/_interaction_box_mouse_bindings.py#L141-L154

File: napari/components/_interaction_box_mouse_bindings.py
Comment (MBPhys) [reply napari/components/_interaction_box_mouse_bindings.py]: Unfortunately this work around does not work, because then the affine event of the selected layer does not response. The event is not connected in this case. Nonetheless I have adapted the layer_transform assignment for more readability.

https://github.com/napari/napari/blob/6352c2f634a5ecefb944f2b68395adaf63b54a0f/napari/components/_interaction_box_mouse_bindings.py#L184-L201

File: napari/components/_interaction_box_mouse_bindings.py
Comment (alisterburt) [reply napari/components/_interaction_box_mouse_bindings.py]: this makes sense! agree that it's better not to recompute if it can be avoided

File: napari/components/_interaction_box_mouse_bindings.py
Comment (alisterburt) [reply napari/components/_interaction_box_mouse_bindings.py]: works for me!

**Review by alisterburt — COMMENTED** [review]
this is looking clean now @MBPhys, thanks for the quick turnaround! I'm not sure why but CI appears to have not triggered properly on your merge commit, could you try to push something just to make it kick in so the checks run?

That 'something' could be a simple test which checks that the interaction box is updated when the transform on the layer is updated :)

something like

```python
def test_interaction_box_changes_with_layer_transform(make_napari_viewer):
    viewer = make_napari_viewer()
    image_layer = viewer.add_image(np.random.random((28, 28))
    image_layer.mode = 'transform'
    initial_selection_box = np.copy(viewer.overlays.interaction_box.selection_box_final)
    image_layer.scale = 5
    assert not np.allclose(initial_selection_box, viewer.overlays.interaction_box.selection_box_final)
```

**Comment by MBPhys** [comment]
Good morning @alisterburt,

I have added a test function as you have suggested:) I tested this locally and this works. Hopefully, it works now.
Thank you very much for your help, input and review:) It is up-to-date, it should work:)

**Review by alisterburt — APPROVED** [review]
Perfect, thank you @MBPhys ! I'm approving, we just need approval from another core dev then we can get this merged 🚀

**Comment by MBPhys** [comment]
Okay thanks @alisterburt,

I do not understand why the PR Test for ubuntu, windows and macos fail?
This is strange. I observed this on some other pull requests, too. Do you have any idea, where it comes from?

**Comment by alisterburt** [comment]
@MBPhys I asked @tlambert03 and he suggested maybe the way you were using settings could be the culprit - I've opened a PR at MBPhys/napari#1 so that we can test that.

A more general note (also brought up by @tlambert03) is that we want to avoid dynamically adding internal keybindings where possible, instead we would want to add them to the [global viewer keybindings file](https://github.com/napari/napari/blob/main/napari/components/_viewer_key_bindings.py) and have them only active in a certain [context](https://napari.org/guides/contexts_expressions.html) - this is beyond the scope of this PR as we don't have a context object for the viewer yet, just noting here :)

**Comment by MBPhys** [comment]
Hi @alisterburt,
this could be the case. He is the expert. Thanks for asking:) And thanks for the answering @tlambert03 
Let test that:)

Yes I understand this. I was also confused as I have seen this. For me the  [global viewer keybinding files](https://github.com/napari/napari/blob/main/napari/components/_viewer_key_bindings.py) would be also better, but I did not want to change so much. The structure was already there.
I can make a new commit to put this out of the interaction box file into the global viewer file. Then it is more straightforward and not so confused. What do you think?

**Comment by alisterburt** [comment]
@MBPhys I would say let's keep this PR simple and confined to supporting transforms for the interaction box, that refactor can come later :)

Looking at other active PRs it does seem like the test errors are isolated to this PR. Let's try to simplify things a little bit and make this PR the absolute minimum required for the functionality by
- removing any reference to settings
- leave the keybinding as it was

If this solves the CI issues then we can at least move forward here and address those issues separately, how does that sound?

**Comment by alisterburt** [comment]
waiting on CI now!

**Comment by MBPhys** [comment]
@alisterburt 

This sounds good. Yes, the goal of this PR was to support transform and not to merge keybindings. 
This issue should be addressed separately, but it is a important point:)

So let try:) Keybindings can be confused...

**Comment by MBPhys** [comment]
I'm excited:)

**Comment by alisterburt** [comment]
https://github.com/napari/napari/runs/5692754514?check_suite_focus=true#step:9:211 my bad, I should have thought about this when suggesting you add the test - this test is failing in 'headless' mode (where we run without a qt backend)

This came up recently in #4245 and the solution is that
- tests of napari components should not depend on qt
- we need the viewer for this test, so let's move move the test to `napari._tests/test_interaction_box_with_qt.py` 🙂 

nearly there!

**Comment by tlambert03** [comment]
> we need the viewer for this test,

if ViewerModel is sufficient: that is, if this doesn't require any of the logic in the _qt folder (and it appears not to) then use that instead of make_napari_viewer:

```
from napari.viewer import ViewerModel

def test_something():
    viewer = ViewerModel()
```

**Comment by MBPhys** [comment]
Okay:) Next try:)  @alisterburt 

Thanks @tlambert03 for your help.
I learn more and more about your test architecture:) This is very fancy:)

**Comment by MBPhys** [comment]
@alisterburt 

Sorry this was my fault. I was too fast... with the test function

**Comment by MBPhys** [comment]
@alisterburt  
Hopefully, this should be the last commit on this small bug/ feature added:)

File: napari/components/_tests/test_interaction_box.py
Comment (alisterburt) [inline napari/components/_tests/test_interaction_box.py]: should this be `assert not np.allclose(bla?)` 🙂

**Review by alisterburt — CHANGES_REQUESTED** [review]

File: napari/components/_tests/test_interaction_box.py
Comment (MBPhys) [reply napari/components/_tests/test_interaction_box.py]: You are right. I will make a new push to move them to napari._tests. Then it should be work.
Have a nice weekend

**Review by alisterburt — APPROVED** [review]
Great work here @MBPhys ! I'm sorry we've had issues with the CI here which have made things harder, thank you for holding tight 🙂 

I'm approving because minreq fails are not specific to this PR and the failure seen is partially fixed by #4310 

edit: it would be great if maybe @tlambert03, @kevinyamauchi or @brisvag could give this a once over before merge :)

**Comment by MBPhys** [comment]
Thanks, @alisterburt :) I have followed this:) Yeah this can happen, unfortunately. meshzoo [#4325](https://github.com/napari/napari/pull/4325) could also be the problem, or?
To figure this out, it is hard job. My respect for this great job and the testing procedure.
Hopefully, I can help more in the future concerning testing, Now that I'm more familiar with the napari test architecture:)

Is there any missing for this PR? Should I make a new push, because min_req still fails for the old one?

**Comment by alisterburt** [comment]
@MBPhys I think you've done everything that's needed here, thank you! The remaining issues are separate to this PR so should be resolved separately

Looking forward to seeing you around the repo more! 🙂

File: napari/_tests/test_interactive_transforms.py
Comment (brisvag) [inline napari/_tests/test_interactive_transforms.py]: Could this test be more specific? Something like:

```
assert np.allclose(box_before * 5, box_after)
```

File: napari/components/_interaction_box_mouse_bindings.py
Comment (brisvag) [inline napari/components/_interaction_box_mouse_bindings.py:123]: Is there a reson why this is split up like this, rather than just copying the full layer affine?

**Review by brisvag — APPROVED** [review]
This looks very clean @MBPhys, thanks for the PR! I had a couple of comments, but they're not blocking, so I'm approving.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (MBPhys) [reply napari/components/_interaction_box_mouse_bindings.py:123]: The interaction box should operate on the affine matrix on top of the translation, scale, shear and rotation. Is there any parameter, which holds the composed affine parameters for scale, shear, translation and rotation?

File: napari/_tests/test_interactive_transforms.py
Comment (MBPhys) [reply napari/_tests/test_interactive_transforms.py]: Thanks:) Yes, of course:) I change this, because I need a push to trigger the CI again.

**Comment by alisterburt** [comment]
thanks for the ping @MBPhys - have launched CI :) excited to get this in!! 🎉 

Once tests are passing (or at least all failures are deemed unrelated) we can merge here

**Comment by alisterburt** [comment]
All checks passing! Great work here @MBPhys - this is your first on the main repo right? Congratulations on this first napari PR, excited to see what you'll do next 🚀 🎉

**Comment by MBPhys** [comment]
Thanks, @alisterburt :)
Yes, this is correct. There will be some other PR in the future concerning keybindings and adding a ruler [#4200](https://github.com/napari/napari/issues/4200), but this needs more discussion to ensure that this will be not against some other PR and ideas. Maybe it will be useful to make stuff together so that people do not work past each other. I'm looking forward to see how napari evolves, how I can help :smile:, discuss stuff.
I'm very interested to see what your ideas are for labeling stuff. Perhaps I can help and we can discuss something if you want? 
Thank you very much for your support and reviews @alisterburt, @brisvag, 
in particular to solve the bug with min_req test:) Great job.

**Comment by alisterburt** [comment]
I'll ping you on Zulip - will be great to have a chat sometime!

**Pull request merged**
